Add return values for selected layers in LayersControl

reyemb opened the pull request:
**Description:**

I primarily use st_folium in combination with LayerControl in Streamlit applications. However, one challenge I encountered is that there is no straightforward way to determine which layers are currently selected from the LayerControl. This limits functionality when trying to dynamically update components such as legends for each layer.

**Here's a basic example of the issue:**

  - The user toggles visibility of multiple layers using LayerControl.
  - There is no direct support to return the currently selected layers, which makes it difficult to update elements like legends that correspond to those layers.

To address this issue, I introduced a modification that allows st_folium to return the currently selected layers when interacting with the LayerControl. Now, when users toggle visibility of layers, the Streamlit app can respond accordingly, such as updating a legend for each selected layer. This greatly enhances interactivity.

blackary commented on streamlit_folium/__init__.py:
```diff
@@ -277,7 +278,8 @@ def st_folium(
     # value of the component before the user has interacted with it.
-
+    
+    # Define WMTS layers
```
This comment seems out-of-place

blackary commented on streamlit_folium/frontend/src/index.tsx:
```diff
@@ -111,6 +113,18 @@ function onDraw(e: any) {
 
+function removeLayer(e: any) {
+  const global_data = window.__GLOBAL_DATA__
+  global_data.selected_layers.delete(e.layer["wmsParams"]["layers"]);
```
Does this generically apply to any type of map layer added? Or is this just certain types of layers?

blackary commented:
@reyemb this looks like a great addition to me -- there are a couple of formatting issues that github actions caught. Would you mind fixing those? You can fix those manually, or just run `pre-commit install` and `pre-commit run --all-files` locally to have them auto-fixed for you.

reyemb commented:
Thank you for the feedback!

Here are the changes I made in response to the issues raised:

**Removed the Out-of-Place Comment:**

I removed the comment that was flagged as out-of-place in the previous review.

**Added an Example File:**

I provided an additional example file showcasing how the newly introduced functionality.

Support for All Layer Types:

In response to the feedback about whether the solution applies to different types of map layers: Yes, the solution I've implemented supports all types of layers added to the map, not just specific ones.

Previously, I was only returning the names of the layers, but I found that some providers use layers with identical names while differentiating by URLs. For this reason, the updated implementation now returns a string containing a combination of both the layer's name and URL.
The format is:
  - layer_name1;url1,layer_name2;url2
  - Layers and URLs are separated by ";"
  - Multiple layers are separated by ","

blackary commented:
That way of separating out layers seems OK, but I think it would be better to just return them as json objects, like 
```json
[{"name": "foo", "url": "bar"}...]
```

Is there a good reason not to return them like that?

Also, can you please fix the linting changes? As I suggested here https://github.com/randyzwitch/streamlit-folium/pull/217#issuecomment-2377664168, you can do this locally with pre-commit.

reyemb commented:
Hey I thought I might go the easy way, but I adjusted to code and run pre-commit as you suggested.

reyemb commented:
I dont know why the test fails. It worked nicely on my side :(

<?xml version="1.0" encoding="utf-8"?>
<testsuites>
	<testsuite name="pytest" errors="0" failures="0" skipped="0" tests="22" time="58.855" timestamp="2024-09-27T17:29:11.061097+02:00" hostname="D01-EAS-V-NB208">
		<testcase classname="tests.test_frontend" name="test_marker_click[chromium]" time="10.365" />
		<testcase classname="tests.test_frontend" name="test_draw[chromium]" time="2.962" />
		<testcase classname="tests.test_frontend" name="test_limit_data[chromium]" time="2.727" />
		<testcase classname="tests.test_frontend" name="test_dual_map[chromium]" time="2.487" />
		<testcase classname="tests.test_frontend" name="test_vector_grid[chromium]" time="2.363" />
		<testcase classname="tests.test_frontend" name="test_tooltip_click[chromium]" time="2.775" />
		<testcase classname="tests.test_frontend" name="test_popup_text[chromium]" time="2.828" />
		<testcase classname="tests.test_frontend" name="test_return_on_hover[chromium]" time="3.067" />
		<testcase classname="tests.test_frontend" name="test_responsiveness[chromium]" time="3.901" />
		<testcase classname="tests.test_frontend" name="test_geojson_styles[chromium]" time="3.674" />
		<testcase classname="tests.test_frontend" name="test_grouped_layer_control[chromium]" time="2.631" />
		<testcase classname="tests.test_frontend" name="test_geojson_popup[chromium]" time="1.617" />
		<testcase classname="tests.test_frontend" name="test_dynamic_feature_group_update[chromium]" time="4.942" />
		<testcase classname="tests.test_frontend" name="test_layer_control_dynamic_update[chromium]" time="10.786" />
		<testcase classname="tests.test_package" name="test_map" time="1.334" />
		<testcase classname="tests.test_package" name="test_layer_control" time="0.006" />
		<testcase classname="tests.test_package" name="test_draw_support" time="0.007" />
		<testcase classname="tests.test_package" name="test_map_id" time="0.003" />
		<testcase classname="tests.test_package" name="test_feature_group" time="0.004" />
		<testcase classname="tests.test_package" name="test_dual_map" time="0.011" />
		<testcase classname="tests.test_package" name="test_vector_grid" time="0.003" />
		<testcase classname="tests.test_release" name="test_release" time="0.000" />
	</testsuite>
</testsuites>

randyzwitch commented:
Looks like the tests are being flaky @reyemb with a bunch of timeouts happening. Me or @blackary will try running the tests locally and if they pass, then we'll merge

randyzwitch commented:
All but one passes locally for me @reyemb

```
============================================================= FAILURES =============================================================
___________________________________________ test_layer_control_dynamic_update[chromium] ____________________________________________

page = <Page url='http://localhost:8699/dynamic_layer_control'>

    def test_layer_control_dynamic_update(page: Page):
        page.get_by_role("link", name="dynamic layer control").click()
        # page.get_by_text("Show generated code").click()
    
        page.frame_locator('iframe[title="streamlit_folium\\.st_folium"]').get_by_text(
            "Parcels"
        ).click()
        expect(
            page.frame_locator('iframe[title="streamlit_folium\\.st_folium"]').get_by_text(
                "Parcels"
            )
        ).not_to_be_checked()
        expect(
            page.frame_locator('iframe[title="streamlit_folium\\.st_folium"]')
            .get_by_role("img")
            .locator("svg")
        ).to_be_hidden()
        expect(page.get_by_text("dashArray")).to_be_hidden()
    
        page.get_by_test_id("stRadio").get_by_text("Both").click()
        page.frame_locator('iframe[title="streamlit_folium\\.st_folium"]').get_by_text(
            "Parcels"
        ).click()
        expect(
            page.frame_locator('iframe[title="streamlit_folium\\.st_folium"]').get_by_text(
                "Parcels"
            )
>       ).not_to_be_checked()
E       AssertionError: Locator expected not to be checked
E       Actual value: True 
E       Call log:
E       LocatorAssertions.not_to_be_checked with timeout 5000ms
E         - waiting for frame_locator("iframe[title=\"streamlit_folium\\.st_folium\"]").get_by_text("Parcels")
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"
E         -   locator resolved to <span> Parcels</span>
E         -   unexpected value "checked"

tests/test_frontend.py:384: AssertionError
========================================================= warnings summary =========================================================
tests/test_package.py::test_map
  <frozen importlib._bootstrap>:488: DeprecationWarning: Type google.protobuf.pyext._message.ScalarMapContainer uses PyType_Spec with a metaclass that has custom tp_new. This is deprecated and will no longer be allowed in Python 3.14.

tests/test_package.py::test_map
  <frozen importlib._bootstrap>:488: DeprecationWarning: Type google.protobuf.pyext._message.MessageMapContainer uses PyType_Spec with a metaclass that has custom tp_new. This is deprecated and will no longer be allowed in Python 3.14.

tests/test_package.py::test_map
  /Users/randyzwitch/miniconda3/envs/streamlit-folium/lib/python3.12/site-packages/google/protobuf/internal/well_known_types.py:91: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    _EPOCH_DATETIME_NAIVE = datetime.datetime.utcfromtimestamp(0)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------- generated xml file: /Users/randyzwitch/github/streamlit-folium/test-results.xml --------------------------
===================================================== short test summary info ======================================================
FAILED tests/test_frontend.py::test_layer_control_dynamic_update[chromium] - AssertionError: Locator expected not to be checked
============================================ 1 failed, 21 passed, 3 warnings in 35.83s =============================================
(streamlit-folium) randyzwitch@Randys-MBP streamlit-folium % 
```

randyzwitch commented:
Going to merge this. I believe that #219 from @blackary might fix the test, not sure

randyzwitch commented:
@reyemb ...my apologies, but can you re-submit this PR? I made a logical error about the order I could merge these PRs, and I cannot re-open this one

Should just be hitting the button again from your fork to re-submit
